File: sdks/java/integration-tests/src/test/java/inc/reactor/sdk/integration/LiveMediaTest.java
package inc.reactor.sdk.integration;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import inc.reactor.sdk.AudioFrame;
import inc.reactor.sdk.PublishState;
import inc.reactor.sdk.ReactorException;
import inc.reactor.sdk.Track;
import inc.reactor.sdk.VideoFrame;
import java.time.Duration;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.Timeout;

/**
 * Media, both directions, over a real connection.
 *
 * <p>{@code reactor/echo} sends back what it is given, so the send path can be checked by what
 * comes out rather than by the absence of an error — which is the whole difficulty with a
 * permissive native layer.
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
final class LiveMediaTest extends LiveFixture {

    private static final int WIDTH = 320;
    private static final int HEIGHT = 240;

    @Test
    @DisplayName("published frames come back out the other side")
    @Timeout(240)
    void publishedFramesComeBack() throws Exception {
        Track input = shared.track(Live.VIDEO_IN);
        Track output = shared.track(Live.VIDEO_OUT);

        AtomicInteger received = new AtomicInteger();
        AtomicReference<String> size = new AtomicReference<>();
        var subscription = output.onFrame((VideoFrame frame) -> {
            received.incrementAndGet();
            size.compareAndSet(null, frame.width() + "x" + frame.height());
        });

        assertEquals(PublishState.UNPUBLISHED, input.publishState());
        input.publish().join();
        assertEquals(PublishState.PUBLISHED, input.publishState());

        // Pushed on a loop rather than once: a single frame can be swallowed by a pipeline still
        // warming up, and this is checking that the path works rather than that every frame
        // survives it.
        for (int index = 0; index < 150 && received.get() < 5; index++) {
            input.pushFrame(bar(index), WIDTH, HEIGHT);
            Thread.sleep(33);
        }

        Live.await("frames to come back from echo", Duration.ofSeconds(60), () -> received.get() > 0);
        System.out.println("received " + received.get() + " frames at " + size.get());

        input.unpublish();
        assertEquals(PublishState.UNPUBLISHED, input.publishState());
        subscription.close();
    }

    @Test
    @DisplayName("pushing before publishing raises against a real session, rather than vanishing")
    @Timeout(180)
    void pushingBeforePublishingRaises() {
        Track input = shared.track(Live.VIDEO_IN);
        if (input.publishState() == PublishState.PUBLISHED) {
            input.unpublish();
        }

        // The native layer would take this frame and drop it. Against a live session that is
        // indistinguishable from a model ignoring you, which is why it raises here.
        ReactorException thrown = assertThrows(ReactorException.class, () -> input.pushFrame(bar(0), WIDTH, HEIGHT));

        assertEquals("INVALID_STATE", thrown.code());
        assertTrue(thrown.getMessage().contains("publish()"), thrown.getMessage());
    }

    @Test
    @DisplayName("receiving on a sendonly track raises, rather than never firing")
    @Timeout(180)
    void receivingOnASendonlyTrackRaises() {
        ReactorException thrown = assertThrows(
                ReactorException.class, () -> shared.track(Live.VIDEO_IN).onFrame((VideoFrame frame) -> {}));

        assertTrue(thrown.getMessage().contains("recvonly"), thrown.getMessage());
    }

    @Test
    @DisplayName("audio travels too, and arrives as PCM the core's contract describes")
    @Timeout(240)
    void audioTravels() throws Exception {
        Track microphone = shared.track(Live.AUDIO_IN);
        Track speaker = shared.track(Live.AUDIO_OUT);

        AtomicInteger received = new AtomicInteger();
        AtomicReference<String> format = new AtomicReference<>();
        var subscription = speaker.onFrame((AudioFrame frame) -> {
            received.incrementAndGet();
            format.compareAndSet(null, frame.sampleRate() + " Hz / " + frame.channels() + " ch");
        });

        microphone.publish().join();

        // 20 ms of silence per push, at 48 kHz mono: the frame size every Reactor model takes.
        short[] silence = new short[48_000 / 50];
        for (int index = 0; index < 150 && received.get() < 5; index++) {
            microphone.pushFrame(silence, 48_000, 1);
            Thread.sleep(20);
        }

        Live.await("audio to come back from echo", Duration.ofSeconds(60), () -> received.get() > 0);
        System.out.println("received " + received.get() + " audio frames at " + format.get());

        microphone.unpublish();
        subscription.close();
    }

    @Test
    @DisplayName("pause and resume take effect on a live track")
    @Timeout(240)
    void pauseAndResumeTakeEffect() {
        Track output = shared.track(Live.VIDEO_OUT);

        output.pause().join();
        Live.await("the pause to be reported", Duration.ofSeconds(30), output::isPaused);

        output.resume().join();
        Live.await("the resume to be reported", Duration.ofSeconds(30), () -> !output.isPaused());
    }

    /** A frame with something in it that moves, so an echo is visibly an echo. */
    private static byte[] bar(int index) {
        byte[] bgra = new byte[WIDTH * HEIGHT * 4];
        int barX = index * 5 % WIDTH;
        for (int y = 0; y < HEIGHT; y++) {
            for (int x = 0; x < WIDTH; x++) {
                int at = (y * WIDTH + x) * 4;
                boolean onBar = Math.abs(x - barX) < 16;
                bgra[at] = (byte) (onBar ? 220 : 20);
                bgra[at + 1] = (byte) (onBar ? 180 : 40);
                bgra[at + 2] = (byte) (onBar ? 60 : 90);
                bgra[at + 3] = (byte) 255;
            }
        }
        return bgra;
    }
}

File: sdks/android/integration-tests/src/androidTest/kotlin/inc/reactor/sdk/android/integration/LiveMediaTest.kt
package inc.reactor.sdk.android.integration

import androidx.test.ext.junit.runners.AndroidJUnit4
import inc.reactor.sdk.android.AudioFrame
import inc.reactor.sdk.android.InvalidStateException
import inc.reactor.sdk.android.VideoFrame
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.FixMethodOrder
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.MethodSorters
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

/**
 * Media over the wire, both directions.
 *
 * Echo returns what it is sent, which is what makes a round trip assertable at all: a frame
 * pushed into `webcam` comes back on `main_video`. Nothing else published does that.
 */
@RunWith(AndroidJUnit4::class)
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
internal class LiveMediaTest : LiveFixture() {
    private companion object {
        const val WIDTH = 64
        const val HEIGHT = 64
    }

    /**
     * 10ms of interleaved signed 16-bit PCM at 48kHz mono, direct because the native layer reads
     * it in place. 480 samples *per channel* — the ABI's units, not the total.
     */
    private fun pcm(): ByteBuffer {
        val buffer = ByteBuffer.allocateDirect(480 * 2).order(ByteOrder.nativeOrder())
        repeat(480) { buffer.putShort((it * 64).toShort()) }
        buffer.rewind()
        return buffer
    }

    /** A solid-colour BGRA frame, tagged so the frame that comes back is identifiably ours. */
    private fun frame(tag: Byte): ByteBuffer {
        val pixels = ByteBuffer.allocateDirect(WIDTH * HEIGHT * 4).order(ByteOrder.nativeOrder())
        repeat(WIDTH * HEIGHT) {
            pixels.put(tag) // B
            pixels.put(0x20) // G
            pixels.put(0x40) // R
            pixels.put(0xFF.toByte()) // A
        }
        pixels.rewind()
        return pixels
    }

    @Test
    fun publishedFramesComeBack() =
        runBlocking {
            val outgoing = reactor.track(Live.VIDEO_IN)
            val incoming = reactor.track(Live.VIDEO_OUT)
            val received = AtomicInteger()
            val shape = AtomicReference<String>()

            incoming.onFrame { frame: VideoFrame ->
                received.incrementAndGet()
                shape.compareAndSet(null, "${frame.width}x${frame.height}")
            }
            outgoing.publish()

            try {
                // Pushed repeatedly rather than once: the first frames of a freshly negotiated
                // track are routinely dropped while the encoder settles, and a single push would
                // make this test a coin flip.
                Live.await("a pushed frame to come back", timeoutMs = 60_000) {
                    outgoing.pushFrame(frame(0x7F), WIDTH, HEIGHT)
                    received.get() > 0
                }
                assertNotNull("the returned frame must carry its shape", shape.get())
            } finally {
                incoming.removeHandlers()
                runCatching { outgoing.unpublish() }
            }
        }

    /**
     * Pushing before publishing is refused rather than dropped.
     *
     * The FFI takes the frame and discards it — there is no sender behind the slot yet — so the
     * caller sees a loop pushing at 30fps and a model receiving nothing. This is the invariant
     * that failure mode exists to prevent, checked against a real session.
     */
    @Test
    fun pushingBeforePublishingIsRefused() {
        val outgoing = reactor.track(Live.AUDIO_IN)
        assertTrue("this test needs an unpublished track", !outgoing.published)
        assertThrows(InvalidStateException::class.java) {
            outgoing.pushFrame(pcm(), samplesPerChannel = 480, sampleRate = 48_000, channels = 1)
        }
    }

    @Test
    fun receivingOnASendonlyTrackIsRefused() {
        assertThrows(InvalidStateException::class.java) {
            reactor.track(Live.VIDEO_IN).onFrame { _: VideoFrame -> }
        }
    }

    @Test
    fun pushingIntoARecvonlyTrackIsRefused() {
        assertThrows(InvalidStateException::class.java) {
            reactor.track(Live.VIDEO_OUT).pushFrame(frame(0x01), WIDTH, HEIGHT)
        }
    }

    @Test
    fun audioTravels() =
        runBlocking {
            val outgoing = reactor.track(Live.AUDIO_IN)
            val incoming = reactor.track(Live.AUDIO_OUT)
            val received = AtomicInteger()
            val rate = AtomicInteger()

            // The parameter is typed because onFrame is overloaded on it — one frame API for both
            // kinds, as every Reactor SDK has — and the kind is not inferable from the body.
            incoming.onFrame { frame: AudioFrame ->
                received.incrementAndGet()
                rate.compareAndSet(0, frame.sampleRate)
            }
            outgoing.publish()

            try {
                Live.await("audio to come back", timeoutMs = 60_000) {
                    outgoing.pushFrame(pcm(), samplesPerChannel = 480, sampleRate = 48_000, channels = 1)
                    received.get() > 0
                }
                assertTrue("the returned audio must declare a sample rate", rate.get() > 0)
            } finally {
                incoming.removeHandlers()
                runCatching { outgoing.unpublish() }
            }
        }

    /**
     * Pause stops delivery and resume restores it.
     *
     * Counted across a window rather than asserted on the next frame: "paused" is a request that
     * takes effect when the platform acts on it, and the frame already in flight when it was sent
     * is not a bug.
     */
    @Test
    fun pauseAndResumeTakeEffect() =
        runBlocking {
            val incoming = reactor.track(Live.VIDEO_OUT)
            val outgoing = reactor.track(Live.VIDEO_IN)
            val received = AtomicInteger()
            incoming.onFrame { _: VideoFrame -> received.incrementAndGet() }
            outgoing.publish()

            try {
                Live.await("frames before pausing", timeoutMs = 60_000) {
                    outgoing.pushFrame(frame(0x11), WIDTH, HEIGHT)
                    received.get() > 0
                }

                incoming.pause()
                assertTrue("the track must report itself paused", incoming.paused)
                kotlinx.coroutines.delay(2_000) // let anything in flight land
                val whilePaused = received.get()
                repeat(20) {
                    outgoing.pushFrame(frame(0x22), WIDTH, HEIGHT)
                    kotlinx.coroutines.delay(50)
                }
                assertEquals(
                    "a paused track must deliver nothing",
                    whilePaused,
                    received.get(),
                )

                incoming.resume()
                assertTrue("the track must report itself resumed", !incoming.paused)
                Live.await("frames after resuming", timeoutMs = 60_000) {
                    outgoing.pushFrame(frame(0x33), WIDTH, HEIGHT)
                    received.get() > whilePaused
                }
            } finally {
                incoming.removeHandlers()
                runCatching { if (incoming.paused) incoming.resume() }
                runCatching { outgoing.unpublish() }
            }
        }
}
